mozetl/sync/bookmark_validation.py

```diff
@@ -0,0 +1,133 @@
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def extract(spark, path, start_date):
+    """Register a temporary `sync_summary` view on the start date."""
+
+    (
+        spark.read.parquet(path)
+        .where(F.col("submission_date_s3") == start_date)
+        .createOrReplaceTempView("sync_summary")
+    )
+
+
+def transform(spark):
+    """Create the bookmark problem and summary tables."""
+
+    query = """
+    SELECT s.app_build_id, s.app_version, s.app_display_version, s.app_name,
+       s.app_channel, s.uid, s.deviceID as device_id,
+       s.submission_date_s3 AS submission_day,
+       date_format(from_unixtime(s.when / 1000), 'YYYYMMdd') AS sync_day,
+       s.when,
+       s.status,
+       e.name AS engine_name,
+       e.status AS engine_status,
+       e.failureReason AS engine_failure_reason,
+       e.validation.problems IS NOT NULL AS engine_has_problems,
+       e.validation.version AS engine_validation_version,
+       e.validation.checked AS engine_validation_checked,
+       e.validation.took AS engine_validation_took,
+       p.name AS engine_validation_problem_name,
+       p.count AS engine_validation_problem_count
+    FROM sync_summary s
+    LATERAL VIEW explode(s.engines) AS e
+    LATERAL VIEW OUTER explode(e.validation.problems) AS p
+    WHERE s.failureReason IS NULL
+    """
+    engine_validation_results = spark.sql(query)
+
+    bmk_validation_results = (
+        engine_validation_results
+        .filter(engine_validation_results["engine_name"] == "bookmarks")
+    )
+
+    # Bookmark validations with problems.
+    bmk_validation_problems = (
+        bmk_validation_results
+        .filter(bmk_validation_results["engine_has_problems"])
+    )
+
+    # All bookmark validations, including without problems.
+    bmk_total_per_day = (
+        bmk_validation_results
+        .filter(bmk_validation_results["engine_validation_checked"].isNotNull())
+        .groupBy("sync_day")
+        .agg(
+            F.countDistinct("uid", "device_id", "when").alias("total_bookmark_validations"),
+            F.countDistinct("uid").alias("total_validated_users"),
+            F.sum("engine_validation_checked").alias("total_bookmarks_checked")
+        )
+    )
+
+    bmk_validation_problems.createOrReplaceTempView("bmk_validation_problems")
+    bmk_total_per_day.createOrReplaceTempView("bmk_total_per_day")
+
+
+def load(spark, bucket, prefix, version, start_date):
+    """Save tables to disk."""
+
+    for table_name in ["bmk_validation_problems", "bmk_total_per_day"]:
+        path = (
+            "s3://{}/{}/{}/v{}/start_date={}"
+            .format(bucket, prefix, table_name, version, start_date)
+        )
+
+        logger.info("Saving table {} on start_date {} to {}"
+                    .format(table_name, start_date, path))
+
+        df = spark.sql("SELECT * FROM {}".format(table_name))
+        df.repartition(1).write.parquet(path, mode="overwrite")
```

Comment:
Do you know approximately how large the resulting parquet files will be with a single partition per `start_date`?

Reply:
For `bmk_total_per_day` it's about 1kb. For `sync_validation_problems` it's 2.5kb.